File: 09_Trees/BST/BST.py
from collections import deque

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
 
def insert(node, key):
    if node is None:
        return Node(key)
 
    if key < node.key:
        node.left = insert(node.left, key)
    elif key > node.key:
        node.right = insert(node.right, key)
 
    return node
 
def search(root, key):
    if root is None or root.key == key:
        return root
 
    if root.key < key:
        return search(root.right, key)
 
    return search(root.left, key)
 
 
def print_tree(root):
    if not root:
        return

    queue = deque([(root, 0)])
    prev_level = 0
    while queue:
        node, level = queue.popleft()
        if level != prev_level:
            print()
            prev_level = level
        print(" " * (2 ** (5 - level) - 1), node.key, end=" " * (2 ** (5 - level) - 1))
        if node.left:
            queue.append((node.left, level + 1))
        if node.right:
            queue.append((node.right, level + 1))
    print()
    
if __name__ == '__main__':
    root = None
    root = insert(root, 50)
    insert(root, 30)
    insert(root, 20)
    insert(root, 40)
    insert(root, 70)
    insert(root, 60)
    insert(root, 80)
 
    key = 6
 
    if search(root, key) is None:
        print(key, "not found")
    else:
        print(key, "found")
 
    key = 60
 
    if search(root, key) is None:
        print(key, "not found")
    else:
        print(key, "found")
    print_tree(root)

File: 09_Trees/BST/BST.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <utility>
#include <cmath>

using namespace std;

class Node {
public:
    int key;
    Node *left, *right;

    Node(int item) {
        key = item;
        left = right = nullptr;
    }
};

class BinarySearchTree {
public:
    Node* root;

    BinarySearchTree() {
        root = nullptr;
    }

    Node* insert(Node* node, int key) {
        if (node == nullptr) {
            return new Node(key);
        }

        if (key < node->key) {
            node->left = insert(node->left, key);
        } else if (key > node->key) {
            node->right = insert(node->right, key);
        }

        return node;
    }

    Node* search(Node* root, int key) {
        if (root == nullptr || root->key == key) {
            return root;
        }

        if (root->key < key) {
            return search(root->right, key);
        }

        return search(root->left, key);
    }

    void printTree(Node* root) {
    if (root == nullptr)
        return;
    
    queue<pair<Node*, int>> q;
    q.push({root, 0});
    int prevLevel = 0;
    
    while (!q.empty()) {
        auto [node, level] = q.front();
        q.pop();
        
        if (level != prevLevel) {
            cout << endl;
            prevLevel = level;
        }
        
        int spaces = pow(2, 5 - level) - 1;
        cout << string(spaces, ' ') << node->key << string(spaces, ' ');
        
        if (node->left != nullptr)
            q.push({node->left, level + 1});
        if (node->right != nullptr)
            q.push({node->right, level + 1});
    }
    cout << endl;
}
};

int main() {
    BinarySearchTree tree;
    Node* root = nullptr;
    root = tree.insert(root, 50);
    tree.insert(root, 30);
    tree.insert(root, 20);
    tree.insert(root, 40);
    tree.insert(root, 70);
    tree.insert(root, 60);
    tree.insert(root, 80);

    int key = 6;

    if (tree.search(root, key) == nullptr) {
        cout << key << " not found" << endl;
    } else {
        cout << key << " found" << endl;
    }

    key = 60;

    if (tree.search(root, key) == nullptr) {
        cout << key << " not found" << endl;
    } else {
        cout << key << " found" << endl;
    }

    tree.printTree(root);
    return 0;
}
